### ISSUE TITLE
Spellchecking integration

### ISSUE THREAD

**Issue by the-mikedavis** [helix-editor/helix#11660]
I've switched https://github.com/helix-editor/spellbook (Hunspell-like spellchecking library) to public and published a v0.1.0 release. It's still a work-in-progress but it's matured to the point where the `check` interface works robustly for at least `en_US`. Other languages should work too but there may be bugs with more advanced rules dictionaries can use. `suggest` functionality is what I'm working on now but now that `check` is in place it's a good time to start thinking about the integration with Helix which is also a potentially large amount of work.

I'm using it in my driver branch in ([commit](https://github.com/the-mikedavis/helix/commit/00d2dd2b1b9de9c269d6e9eb43d6c26040518b63)) where the integration is very naive: it checks all words in the viewport render and highlights misspelled words.

<img src="https://github.com/user-attachments/assets/0264e59f-2280-4b3b-984c-c5ef411b49f2" width="75%" />

@pascalkuthe and I have chatted about this and here's what we're thinking for a proper integration:

* Load dictionaries off the main thread: `spellbook::Dictionary::new` can take tens of milliseconds depending on the size of a dictionary.
* Read dictionary files (like you might find at https://github.com/LibreOffice/dictionaries) from a runtime directory. We could try to find system-installed dictionaries to be a bit fancier but that can be done as a follow-up if at all.
* Spellcheck open documents also off the main thread since we're checking the full document.
* Use tree-sitter queries to tag nodes which should be checked. It might also be useful to specify a tokenization strategy when tagging nodes: one mode for prose and another for programming languages.
* When a document changes, recheck incrementally: only recheck around and within the change's range.

Interacting with the spellchecker should be similar to interacting with LSP code-actions. `<space>a` should pull up LSP code actions and actions from the spellchecker: adding a misspelled word to a personal dictionary and suggestions from `suggest`. Adding to a personal dictionary can use the `spellbook::Dictionary::add` function and append a line with the word to a file in `$XDG_STATE_HOME/helix/personal-dictionary/<locale>`.

**Comment by sbromberger** [comment]
Not that I'm against this idea at all - I'm really happy that we're considering adding spell checking as a first class feature - but what would be the advantage of using an integrated spell checker over an LSP-based solution? I've been using vale now for a while and (aside from the fact that it doesn't allow you to add to local dictionaries) it's pretty darn good. I am similarily enthusiastic about Harper, but it appears to have some different limitations.

**Comment by kirawi** [comment]
You would be able to leverage tree-sitter for grammar checking in programming languages.

**Comment by sbromberger** [comment]
> You would be able to leverage tree-sitter for grammar checking in programming languages.

What about natural languages (markdown / text / LaTeX?)

**Comment by kirawi** [comment]
It should work as normal with LSPs now.

**Comment by sbromberger** [comment]
Then I guess I'm not seeing as much of an advantage if I have to use an LSP in any case to get grammar checking. Most grammar checkers (vale, harper) do spell checking as well....

**Comment by kirawi** [comment]
Yeah, it's most beneficial for programming languages.

**Comment by summersz** [comment]
Looking forward to this. I have used CSpell in the past so having a different strategy/config for code and prose sounds like a good idea and is a limitation of the lsp alternatives.

**Comment by pascalkuthe** [comment]
Yeah helix having a universally parser/sytanx tree woth tree sitter that we can use for spellchecker is something an LSP can't really match.

Spellchecking is also a fairly basic/common feature that is nice to have on core withlut having to use/setup a third party LSP.

**Comment by justinlovinger** [comment]
Vale and Harper can already spellcheck comments in code. `typos` can additionally spellcheck variable-names.

**Comment by justinlovinger** [comment]
I have concerns this feature will not age well and will eventually become a "thing new users should turn off and replace". Linting natural language is incredibly complicated, and even the best solutions today are far from ideal. I think spelling and grammar checking should be left to integration with external tools, not a part of Helix itself.

**Comment by the-mikedavis** [comment]
Nothing's stopping you from continuing to use something else via LSP if you choose. You shouldn't need a language server installed for basic spellchecking (word checking, not grammar) though.

In any case this issue is about the technical aspects of integration. Discussion on the concept of integrating spellchecking is in https://github.com/helix-editor/helix/discussions/3637

**Comment by jsco2t** [comment]
I'm a relatively new user to Helix - so take that as the *important* context in this statement. I spent a non-trivial bit of time setting up LSP's (in my case Vale) just so that I could get spell-checking in Markdown docs. 

I get the concerns about the feature potentially not aging well. I think the simple solution to that (similar to what vim/nvim does) is to allow the built-in spell-checker to be turned off if needed.

For me - having a built-in spellchecker would have been a productivity/usability boost as a new user (a user who also sucks at spelling occasionally)

**Comment by pascalkuthe** [comment]
I am unconvinced by `typos`(-ls) approach. They aim for low false positives rate and only check for known typos instead of checking words against dictonaries which is why they can check programming languages without parsing the sourccode. However I find the false negative rate way too high and find the results with traditional spell checkers much better.

With helix we are in a position where we do have a universal parser for every language and that means we can enable a much more reliable spellchecker. This is something that is impractical (and inefficient) for a LS to replicate.

**Comment by hnorkowski** [comment]
> I'm using it in my driver branch in ([commit](https://github.com/the-mikedavis/helix/commit/00d2dd2b1b9de9c269d6e9eb43d6c26040518b63)) where the integration is very naive: it checks all words in the viewport render and highlights misspelled words.

 I wanted to try it but there is a private dependency in it:
```
skidder = { git = "https://github.com/helix-editor/tree-sitter-helix.git" }
```

**Comment by the-mikedavis** [comment]
Yeah that commit is on my driver branch which has a bunch of other changes. I pushed up another [branch](https://github.com/helix-editor/helix/tree/spellbook) off of master: [commit](https://github.com/helix-editor/helix/commit/794b2bef345dc7474013b1810857aa058434cac1). You'll need a dictionary downloaded into `$HELIX_RUNTIME/dictionaries/en_US/en_US.{dic,aff}` to run that commit. You can download one from http://wordlist.aspell.net/dicts/ or copy the right files out of https://github.com/LibreOffice/dictionaries or https://searchfox.org/mozilla-central/source/extensions/spellcheck/locales/en-US/hunspell/dictionary-sources/utf8

**Comment by hnorkowski** [comment]
@the-mikedavis Is it possible to underline the spelling issues in a different color than the LSP issues?

**Comment by the-mikedavis** [comment]
I was imagining that there would be diagnostics (same as LSP diagnostics) for misspelled words. They could be at a different or maybe customizable severity which would affect how they're displayed. In the linked commit/branch I'm using `diagnostic.error` only for simplicity

**Comment by mrnossiom** [comment]
I don't know if you've already started this implementation. I wanted to do something similar last year as an LSP for spellchecking support called `lspelling`. (Also made a custom incomplete hunspell port as the spelling checker engine called `ruspell`). I had arrived to the same conclusion as the one listed in the initial message.

In the [wordc](https://github.com/mrnossiom/lspelling/blob/main/crates/wordc/src/lib.rs) part of `lspelling`, I would parse sources files with TS, fragment the relevant bits with a [TS query](https://github.com/mrnossiom/lspelling/blob/main/crates/wordc/queries/rust.scm) keeping prose distinct, process fragments with the chosen spellchecking backend (e.g. spellbook).

I would be happy to help on this feature

**Comment by cgahr** [comment]
The spellbook branch looks already really usable to me. Highlight spelling mistakes alone is a game changer for me.
> You can download one from http://wordlist.aspell.net/dicts/ or copy the right files out of https://github.com/LibreOffice/dictionaries or https://searchfox.org/mozilla-central/source/extensions/spellcheck/locales/en-US/hunspell/dictionary-sources/utf8

Expanding on this a bit:
- you can use http://app.aspell.net/lookup?dict=en_US to look up if your "words" are in the dictionary
- you can create custom dictionaries here: http://app.aspell.net/create?defaults=en_US-large

Is it possible to configure which treesitter objects (?) are spellchecked? For example, what if I only want spellchecking of i.e. variable names?

**Comment by the-mikedavis** [comment]
I believe that branch should be creating the personal dictionary file for you if you use the `add_word_to_personal_dictionary` command. I should update that branch now that Spellbook has suggestions support - you should be able to get a popup of suggestions (same UI as code actions) for spelling errors.

That branch is really hacky though - there isn't any interaction with tree-sitter to (dis)qualify nodes to check yet and the checking is run per-render on the visible viewport.

**Comment by cgahr** [comment]
> I believe that branch should be creating the personal dictionary file for you if you use the `add_word_to_personal_dictionary` command. 
Yes and no: if the directory `~/.local/state/helix` exists, the file is created automatically. If the directory does not exist, `add_word_to_personal_dict` fails with an OS error 2 (I think)

> That branch is really hacky though - there isn't any interaction with tree-sitter to (dis)qualify nodes to check yet and the checking is run per-render on the visible viewport.
Yes, I am aware! However, and that's kudos to you, it already works really and is much more performant than for example https://github.com/ltex-plus/ltex-ls-plus and does what it's supposed to.

From a users perspective, only treesitter is missing to decrease the number of false-positives.

Great work already!

**Comment by summersz** [comment]
I have just started testing this branch too, and I agree it is working really well.

Is it possible to add words to a dict on a project level?

I have used cSpell in the past and that only checks words that are greater than 3 characters long which removes a lot of false negatives. It would be useful if that could be added as a configurable option. 

If a single word within a kebab-case string is identified as misspelled it highlights the whole string. Not sure if this is intended, but I would expect it just to highlight the misspelled word(s) like it is doing for other cases.

**Comment by cgahr** [comment]
Another feature that would be nice to have are custom affixes. For example, my custom dictionary contains
```
gridpoint/S
hyperparameter/S
invariant/S
```
and uses existing suffixes to define the plural. 

Now I want add words with the prefix `pre`, like
```
precompute/DG
preprocess/DG
```
Of course I can specify them manually (which is probably most of the time sufficient), but it would be nice to able to add a custom `.aff` file, that extends the existing one. This would allow users to write e.g.
```
compute/DGO
process/DGO

```
with the custom prefix
```
PFX O Y 1
PFX O   0     pre         .
```
In the best case, the words are also merged with existing affixes. In the case of `compute`, we would have `compute/ADSGO`.

**Comment by pascalkuthe** [comment]
> This will be a bit off-topic for PR, but relevant to the topic of "false negatives". [`typos`](https://github.com/crate-ci/typos/tree/master) takes a different approach where it matches again common mispellings instead of against a dictionary with words that are spelled correctly. This catches most of the typos, with almost no false positives.

We are aware of typos but intentionally choose a more traditionam dictionary approach because typoes has way too many false negatives. I would rather add to my personal dictionary and use TS for eliminating false negatives.

**Comment by dpc** [comment]
@pascalkuthe Yeah, I was not suggesting that approach in Helix. I think typos shines in CIs to enforce some base level of quality, where false positives are very very annoying. During text editing most people doesn't mind few false positives. Having said that on Helix Matrix channel someone linked https://github.com/tekumara/typos-lsp and I just configured it in Helix for my `.md` files editing. I'll see how well it works in practice.

**Comment by hholst80** [comment]
Having a spellcheck in parity with Vim would require no manual setup to provision a user dictionary. just press zg/zw will add/remove the word under the cursor to the user dictionary. if it does not exist, it will be created. I think Helix should be no less user friendly than that! :+1:

**Comment by cgahr** [comment]
I just switched my helix version to the newest `spellbook` branch and saw the typos can now be rendered as inline diagnostics. Nice!

I was wondering, however, is it possible to configure what type of diagnostic is emitted (error, warn, info, hint)? As it is right with inline diagnostics enabled, they clutter the screen and distract a lot.
As it is right now, inline diagnostics are too distracting to be enabled..

Attached a screenshot of my helix config showing the problem:

![Image](https://github.com/user-attachments/assets/7ec589ee-a332-4828-a539-c3d6857dc292)

**Comment by the-mikedavis** [comment]
Yeah I haven't added any configuration switches yet but it probably makes sense to allow configuring the diagnostic level. The diagnostics are currently very noisy since the entire document is being checked (i.e. not using tree-sitter yet). Even once the branch is using that though it will probably still be quite noisy. We may want to consider typable commands to check and document and clear the diagnostics so that you can briefly opt-in to checking and not have a lot of diagnostics around all the time.

**Comment by cgahr** [comment]
I encountered the following bug on this branch:

1. execute `hx`
2. type some text in initial buffer
3. open new buffer `:n`
4. in new buffer, close all buffers `:bca`

This leads to the following panic:
```
thread 'main' panicked at helix-view/src/editor.rs:1595:19:
called `Option::unwrap()` on a `None` value
stack backtrace:
   0:     0x565198eb9bc9 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::h1d1b4b5fa6f43fbc
   1:     0x565197f86703 - core::fmt::write::ha0383308c4b73ef5
   2:     0x565198ec681f - std::io::Write::write_fmt::h3d049babadfff437
   3:     0x565198eb9a23 - std::sys::backtrace::BacktraceLock::print::h5052bf7310439213
   4:     0x565198ed933f - std::panicking::default_hook::{{closure}}::h5c03afda8dca06e0
   5:     0x565198ed9130 - std::panicking::default_hook::hf34f99039c48f075
   6:     0x565198ed99ba - std::panicking::rust_panic_with_hook::h0da62165dfee2b9c
   7:     0x565198eba3d6 - std::panicking::begin_panic_handler::{{closure}}::h3aa93158d08ca3fc
   8:     0x565198eb9dd9 - std::sys::backtrace::__rust_end_short_backtrace::h59b242cd399f0746
   9:     0x565198ed947d - rust_begin_unwind
  10:     0x565197e90c90 - core::panicking::panic_fmt::h9f331d0a33e30d27
  11:     0x565197e90d1c - core::panicking::panic::h00a158dc53495379
  12:     0x565197e90a29 - core::option::unwrap_failed::h781eb58569f6973b
  13:     0x565198c0b837 - helix_view::editor::Editor::replace_document_in_view::hd99bdb018ac51da4
  14:     0x565198c0e1e9 - helix_view::editor::Editor::close_document::h184c455bc2c8f7bb
  15:     0x5651988725cb - core::iter::traits::collect::default_extend_tuple_b::h871397ee7b434b71
  16:     0x565198526daf - helix_term::commands::typed::buffer_close_by_ids_impl::h1545752e88840679
  17:     0x565198528caa - helix_term::commands::typed::buffer_close_all::h838ff0388bcb4896
  18:     0x565198549a81 - helix_term::commands::typed::execute_command::hec01abb564140fc7
  19:     0x565198548fed - helix_term::commands::typed::execute_command_line::h1ba2cd8f8af24cfc
  20:     0x5651987e81b2 - helix_term::commands::typed::command_mode::{{closure}}::hce4f689d44d5d739
  21:     0x5651987f65d4 - <helix_term::ui::prompt::Prompt as helix_term::compositor::Component>::handle_event::h10fdbe28f32e4dda
  22:     0x565198552aed - helix_term::compositor::Compositor::handle_event::h7c83d33e7ea55550
  23:     0x565198c4e9d4 - hx::main_impl::{{closure}}::h6fd4a1fe0d74f684
  24:     0x565198c4c807 - tokio::runtime::park::CachedParkThread::block_on::he63e7c30b9b38284
  25:     0x565198c848d1 - tokio::runtime::context::runtime::enter_runtime::h51c737c5187e199f
  26:     0x565198cba2d4 - tokio::runtime::runtime::Runtime::block_on::hbfc6dff2ed56c841
  27:     0x565198c90aa2 - hx::main::h37e97e682da37856
  28:     0x565198ccc7e3 - std::sys::backtrace::__rust_begin_short_backtrace::hca336a64fafd5904
  29:     0x565198c7b96d - std::rt::lang_start::{{closure}}::hcf8577a0c1197a9a
  30:     0x565198ed6aa7 - std::rt::lang_start_internal::hf809820c659d07f5
  31:     0x565198c90b95 - main
  32:     0x7faa4fa2a1fe - __libc_start_call_main
  33:     0x7faa4fa2a2b9 - __libc_start_main_alias_2
  34:     0x565197ef8765 - _start
  35:                0x0 - <unknown>
```

**Comment by partiallyordered** [comment]
This might all be out of scope for now..

It would be great to also have some ability to use multiple dictionaries per file, or per project. The problem is that the mapping of (vocabulary) domain to project is many-to-many. For example, I might like to have a "Kubernetes dictionary" or a ["big data dictionary"](https://pixelastic.github.io/pokemonorbigdata/) that I can use in multiple projects. Or an organisation-specific dictionary; for example if I'm working for an org with a set of organisation-specific terms, products, or features, it would be convenient to share one dictionary between projects for that organisation. As a corollary, it would be useful when adding a word to a dictionary to be prompted for the dictionary (or dictionaries) that word should be added to.

**Comment by MadeInShineA** [comment]
Hello, I'm having trouble setting codebook to work properly with my nix home-manager config, has anyone succeeded in doing so ? 

Here is my config 

```
programs.helix = {
  enable = true;
  
  extraPackages = [
    pkgs.simple-completion-language-server
  ];

  settings = {
    theme = "catppuccin_mocha_transparent";
    editor.bufferline = "multiple";
    editor.line-number = "relative";
    editor.inline-diagnostics.cursor-line = "warning";
    editor.inline-diagnostics.other-lines = "warning";
    editor.cursor-shape = {
      normal = "block";
      insert = "bar";
      select = "underline";
    };
  };

  languages = {
    language-servers = {
      codebook = {
        command = "codebook-lsp";
        args = [ "serve" ];
      };
    };

    language = [
      {
        name = "nix";
        auto-format = true;
        formatter.command = lib.getExe pkgs.nixfmt-rfc-style;
        language-servers = [ "nil" ];
      }
      {
        name = "markdown";
        auto-format = true;
        formatter = {
          command = lib.getExe pkgs.deno;
          args = [ "fmt" "-" "--ext" "md" ];
        };
        language-servers = [
          "marksman"
          "codebook"
        ];
        file-types = [ "md" "mdx" "qmd" ];
      }
    ];
  };
};

```

The codebook-lsp command works properly 

`helix --health markdown` doesn't show the codebook lsp only the marksman one

**Comment by jpds** [comment]
@MadeInShineA The only difference I see between your config and mine is that I have `codebook` in my `extraPackages`.

If that doesn't work - try `nix-shell -p codebook` and move the symlinked `~/.config/helix/languages.toml` elsewhere and temporarily add what's in the [codebook README](https://github.com/blopker/codebook#helix) and validate that works.

Failing that, open a discussion topic here.

**Comment by MadeInShineA** [comment]
Thanks for the help!

Turns out that I was using 

`languages.language-servers` instead of `languages.language-server` 

:)
